### ISSUE TITLE
Barracuda Android Build is crashing

### ISSUE THREAD

**Issue by ashwalcs** [Unity-Technologies/ml-agents#2632]
Tried to use barracuda for object detection. For this have loaded a Custom Vision Model converted into barracuda format.

**Comment by chriselion** [comment]
Hi @ashwalcs,
Can you please provide more information about the crash, such as the callstack? Any other information like the device type and Unity version would probably be helpful too.

**Comment by chriselion** [comment]
If you can provide the source neural network file, that would be even better.

**Comment by ashwalcs** [comment]
Hi @chriselion ,
This is the callstack :

UnityException: DestroyBuffer can only be called from the main thread.
10-04 12:25:04.783 4320 4416 E Unity : Constructors and field initializers will be executed from the loading thread when loading a scene.
10-04 12:25:04.783 4320 4416 E Unity : Don't use this function in the constructor or field initializers, instead move initialization code to the Awake or Start function.
10-04 12:25:04.783 4320 4416 E Unity : at UnityEngine.ComputeBuffer.Dispose (System.Boolean disposing) [0x00000] in <00000000000000000000000000000000>:0
10-04 12:25:04.783 4320 4416 E Unity : at UnityEngine.ComputeBuffer.Dispose () [0x00000] in <00000000000000000000000000000000>:0
10-04 12:25:04.783 4320 4416 E Unity : at Barracuda.ComputeTensorData.Dispose () [0x00000] in <00000000000000000000000000000000>:0
10-04 12:25:04.783 4320 4416 E Unity : at Barracuda.Tensor.Dispose () [0x00000] in <00000000000000000000000000000000>:0
10-04 12:25:04.783 4320 4416 E Unity : at Barracuda.Tensor.Finalize () [0x00000] in <00000000000000000000000000000000>:0
10-04 12:25:04.783 4320 4416 E Unity : UnityEngine.DebugLogHandler:LogException(Exception, Object)
10-04 12:25:04.783 4320 4416 E Unity : UnityEngine.Logger:LogException(Exception, Object)
10-04 12:25:04.783 4320 4416 E Unity : UnityEngine.Debug:LogException(Exception)
10-04 12:25:04.783 4320 4416 E Unity : UnityEngine.UnhandledExceptionHandler:<RegisterUECatcher>m__0(Object, UnhandledExceptionEventAr
10-04 12:25:04.785 4320 4416 E Unity : UnityException: GetStackTraceLogType can only be called from the main thread.10-04 12:25:04.785 4320 4416 E Unity : Constructors and field initializers will be executed from the loading thread when loading a scene.
10-04 12:25:04.785 4320 4416 E Unity : Don't use this function in the constructor or field initializers, instead move initialization code to the Awake or Start function.
10-04 12:25:04.785 4320 4416 E Unity : at Barracuda.D.LogWarning (System.Object message) [0x00000] in <00000000000000000000000000000000>:0
10-04 12:25:04.785 4320 4416 E Unity : at Barracuda.ComputeTensorData.Finalize () [0x00000] in <00000000000000000000000000000000>:0
10-04 12:25:04.785 4320 4416 E Unity : UnityEngine.DebugLogHandler:LogException(Exception, Object)
10-04 12:25:04.785 4320 4416 E Unity : UnityEngine.Logger:LogException(Exception, Object)
10-04 12:25:04.785 4320 4416 E Unity : UnityEngine.Debug:LogException(Exception)
10-04 12:25:04.785 4320 4416 E Unity : UnityEngine.UnhandledExceptionHandler:<RegisterUECatcher>m__0(Object, UnhandledExceptionEventArgs)
10-04 12:25:04.785 4320 4416 E Unity : System.UnhandledExceptionEventHandler:Invoke(Object, UnhandledExceptionEventArgs)
10-04 12:25:04.785 4320 4416 E Unity :
10-04 12:25:04.785 4320 4416 E Unity : (Filename: currently not available on il2cpp Line: -1)
10-04 12:25:04.785 4320 4416 E Unity :
10-04 12:25:04.787 4320 4416 W Unity : GarbageCollector disposing of ComputeBuffer. Please use ComputeBuffer.Release() or .Dispose() to manually release the buffer.
10-04 12:25:04.787 4320 4416 W Unity : UnityEngine.DebugLogHandler:LogFormat(LogType, Object, String, Object[])
10-04 12:25:04.787 4320 4416 W Unity : UnityEngine.Logger:Log(LogType, Object)
10-04 12:25:04.787 4320 4416 W Unity : UnityEngine.Debug:LogWarning(Object)
10-04 12:25:04.787 4320 4416 W Unity : UnityEngine.ComputeBuffer:Dispose(Boolean)
10-04 12:25:04.787 4320 4416 W Unity : UnityEngine.ComputeBuffer:Finalize()
10-04 12:25:04.787 4320 4416 W Unity :
10-04 12:25:04.787 4320 4416 W Unity : (Filename: ./Runtime/Export/Debug/Debug.bindings.h Line: 48)
10-04 12:25:04.787 4320 4416 W Unity :

**Comment by surfnerd** [comment]
The finalizer doesn’t run on the main thread, so it looks like that’s why Unity isn’t happy.  Thanks for the extra info!  Very helpful.

**Comment by surfnerd** [comment]
Hi @ashwalcs,
Are you sure that you are disposing off your data structures properly when the scene is closed?  Can you share your environment with us?

**Comment by ashwalcs** [comment]
**Editor Environment** : 
- Unity 2019.1.14f - Windows 10 - 16Gb Ram.

**Testing Environment** : 
- One Plus 7 Pro - Android 9 Pie - 8Gb Ram
- Moto G6 - Android 9 Pie - 3Gb Ram.
                                      
This the Neural Network file - trained for 2 Objects : https://drive.google.com/file/d/1oDPiWjUgvLylk7nIZWK9s5noGThDXjKL/view?usp=sharing

@surfnerd Yes i am disposing the data structures properly and also disposing the worker after every scan. Basically it crashes after every 3 concurrent detections/ scan.

**Comment by mantasp** [comment]
@ashwalcs could you please share code how are you calling Barracuda worker? 
P.S. If you have to run several concurrent executions then you have to create multiple workers.

**Comment by ashwalcs** [comment]
@mantasp I am not running them concurrently. I am running them in series, one after the other. The model is loaded only once and the worker is created on each scan/detection. After completion of detection, worker is disposed off. Only after disposing the worker is another scan/detection initiated.

Here's a snippet :
` public IEnumerator LoadModel()
    {// Called once 
        Debug.Log("Model loaded");
        labels = label.text.Split(new char[] { '\n' }, StringSplitOptions.RemoveEmptyEntries);
        BarracudaModel = ModelLoader.Load(modelfile);
        yield return null;
    }

public IEnumerator RunTF(float[] imgdata)
    {
StartCoroutine(DecodeTexture(m_Texture,` 416, 416, 0, Flip.NONE, (result) =>
            {
                 StartCoroutine(detector.Detect(result, (FinalResult) => { }));
            }));
    }

public IEnumerator Detect(float[] imgData, System.Action<List<BoundingBox>> callback, int numResultsPerClass = 1, float threshold = 0f)
    {
        var inputs = new Dictionary<string, Tensor>();
        var shape = new TensorShape(1, 416, 416, 3);
        inputs[BarracudaModel.inputs[0].name] = new Tensor(shape, imgData);
        worker.ExecuteAndWaitForCompletion(inputs);
        runner = worker.Peek(BarracudaModel.outputs[0]);
        runner.Dispose();
        yield return null;
        callback(results);
    }

public void sendImage() // Onclick of button
    {
        StartCoroutine(detector.createWorker());
   }

    public IEnumerator createWorker()
    {
        Debug.Log("Create Worker");
        worker = BarracudaWorkerFactory.CreateWorker(BarracudaWorkerFactory.Type.Compute, BarracudaModel, true);
        yield return null;
    }

    public void Close() // On click of close
    {
        Debug.Log("Dispose Worker");
        worker?.Dispose();
        runner?.Dispose();
        labels = null;
    } `

**Comment by mantasp** [comment]
@ashwalcs 
- if you need to run network through many pictures, pass them in batches instead of processing one by one. Basically when you allocate tensor specify `batch size` as first tensor shape parameter `var shape = new TensorShape(batch_size, 416, 416, 3);`
- your example shows that you are not disposing all inputs, you have to dispose any tensor you did allocate via `inputs[BarracudaModel.inputs[0].name] = new Tensor(shape, imgData);`. So these tensors are causing errors you see in the logs.
- almost always it is best to call `worker.Execute()` instead of `worker.ExecuteAndWaitForCompletion(inputs);` your code will block anyway when you access output tensor data for the first time. Also note that if you Peek data you don't have to dispose it, but it's valid only until worker is alive and before second call to Execute() on the same worker instance.

**Comment by ashwalcs** [comment]
@mantasp Thank you, disposing all inputs and tensors worked for me.

**Comment by surfnerd** [comment]
Hey @ashwalcs,
We are going to close this issue since it seems like it has been resolved.  Feel free to open another if you run into issues.
